### ISSUE TITLE
Leviathan

### ISSUE THREAD

**Issue by HellDragon-HK** [GreatEmerald/UT3Vehicles#8]
Issues and notes for the Leviathan from in-game testing

UT2004 Draw  - 1.0?
UT3    Draw  - 1.25
UT3Port Draw - 1.0 / NA I don't think this one needs to change honestly

  - [Priority]  
  - [All tires are floating off the ground]  
  - [None of the 5 turrets have collision]  
  - [Turrets kind of become disconnected when they turn]  
  - [Secondary]  
  - [No headlights in front which there are on UT3 but they are small and easy to miss (stock UT2004 Leviathan has headlights but doesn't cast any light)]  
  - [When deployed there are some team colored lights on the back of the cannon on UT3 that are missing on the port]
  - [Main cannon takes about 5 seconds to turn half way around on UT3 and only 2 with the port]
  - [Extra]    
  - [Port projectile isn't nearly the same, on UT3 it fires what looks like tarydium shards]
  - [Looking at the back of the Levi it looks like the bottom bar was meant to have team lights but doesn't in either game and I could swear I've seen more lights on this thing when it's on blue team]
  - [I question that the UT3 Levi actually might have more health than the UT2004 one and for that matter possibly every UT3 vehicle has more life than UT2004, with UT3 not having a numbered health count I can't tell but it just feels like they can take more punishment on UT3]

**Comment by GreatEmerald** [comment]
Not sure if anything can be done about the tires. Perhaps more experimentation is needed though.
Turret collision, attaching properly and projectiles are all missing features.
The headlights you could try to put up yourself.
More lights for another team doesn't sound reasonable, nor too easily implemented.
I have never seen the Leviatan crush other vehicles, maybe because it's rare and hilariously slow.
The health is in the source code of the mod and UT3: http://greatemerald.xmpcommunity.com/uncodex/UT3-15/Source_utgame/utvehicle_leviathan.html#1222 It's correct, though there seems to be some interaction with turret health too.

**Comment by HellDragon-HK** [comment]
The team lights don't emit any kind of noticeable light, they are just glowing shapes for the most part, that really can't be implemented?  If not that's going to hurt the Nightshade and Viper's appearance a lot.

Hmm to me it felt the Levi wasn't any slower than the Goliath or Paladin and yes it does most certainly crush vehicles, an enemy Scorpion (from it's spawn point) is instantly destroyed and so was the Goliath sitting next to it.

I used a youtube video to check the health and 6500 (which is what it has in UT2004) appears correct.

**Comment by GreatEmerald** [comment]
No, I meant the corona lights, usually both teams have the same number of them. Texture-based lights, if they are always on, can be implemented fine. (If they get turned off at some points, it's more annoying because you need to replace textures all the time, though it's also technically doable.)

Interesting, I've never seen it crush vehicles, but like I said it's rare in UT3 and other vehicles are usually taken by the time you roll to them.

**Comment by HellDragon-HK** [comment]
In Serenity it will roll over everything on it's way to the core you should give it a shot and see for yourself

**Comment by Unre-Alex** [comment]
@HellDragon-HK I'm not sure if you still need it, but [this map guide](http://www.mediafire.com/file/f73228d7iiv34xq/UT2004_and_UT3_Map_Guide.zip) will let you see what and where you can find on every map. XSRepSetMutMSUC4 mutator can let you replace vehicles in UT3.

**Comment by HellDragon-HK** [comment]
I'm on PS3 for UT3 testing so I can't really use mutators unless it's been made for PS3
At the moment I can't find a map that has both a Scavenger and Raptor to impact test

**Comment by Unre-Alex** [comment]
Torlan Necris

**Comment by HellDragon-HK** [comment]
I checked that and all I had were Vipers and Darkwalkers but I'll give it another look thanks

**Comment by Unre-Alex** [comment]
Raptor is in the red base and scavengers are by the prime nodes (you need to be on the blue team) check the guide I linked you

**Comment by HellDragon-HK** [comment]
Sorry at the time I was trying to knock a Nemesis around with the Raptor, still got the Darkwalker to try that with and then I can go check Torlan

**Comment by Unre-Alex** [comment]
let me check

**Comment by HellDragon-HK** [comment]
I gave up on it, I took every node and all I had were Vipers and Darkwalkers....could be a PS3 limitation or only in Campaign if you can I'd appreciate the help filling in the blanks

**Comment by Unre-Alex** [comment]
coldharbor

**Comment by HellDragon-HK** [comment]
Ah right I remember that now, thanks...assuming PS3 isn't different that is

**Comment by Unre-Alex** [comment]
there's no map with leviathan and cicada

**Comment by HellDragon-HK** [comment]
Why does z translation not do anything with the Leviathan, it's still floating and doesn't appear to be changed by any number whatsoever :S

**Comment by HellDragon-HK** [comment]
Driver turret is too small, changing the scale to X = 1.7  Y = 1.55  Z = 1.5 looks about right, @Unre-Alex Hey Alex, if you're around want to see what you get for this one too?

**Comment by GreatEmerald** [comment]
Again, Z translation changes the centre of mass.

**Comment by HellDragon-HK** [comment]
Levi is getting stuck very easily on Urban and not very useful, can anything be done about it, the stock Levi doesn't have those issues does it?

**Comment by Unre-Alex** [comment]
it's wider, Y 1.7

**Comment by HellDragon-HK** [comment]
Alright thanks Alex, I came close at least this time right?   Now if only we could figure out how to put it in code instead.

**Comment by HellDragon-HK** [comment]
@GreatEmerald @100GPing100 A little too complicated for me this time guys
The Levi driver turret needs...
	KDamageImpulse=1000
	VehicleMomentumScaling=1.5
	VehicleDamageScaling=0.5

MASRocketProjectile extended Rocket Launcher which is not a vehicle weapon though so not sure how to work with that

**Comment by HellDragon-HK** [comment]
Anyone have any objections to changing the Stinger Turret visual to lightning gun, it actually looks closer to what we see in UT3 I think, just make the bolts more of a purple blue instead of white blue and you'll have it

**Comment by HellDragon-HK** [comment]
@GreatEmerald @100GPing100 I think I might know what's going on with the floating Levi issue, it's collision in-game doesn't match it's animations, it's almost as if in-game the Levi itself is smaller than it is in the animation file as the collision is further out than the animation file shows, I'm going to play around with Drawscale and see what happens.

**Comment by HellDragon-HK** [comment]
@Unre-Alex The Shock Turret damage is completely off, I'm getting 11 on everything in UT2004 and in UT3 30 to a Goliath, 31 to a Manta, 34 to a Scorpion, and 37 to a Hellbender....this is with the adding damage scaling in the UT3 code which usually isn't this badly off, what do you get Alex?

**Comment by Unre-Alex** [comment]
all get 11 in UT2004 and 30 in UT3

**Comment by HellDragon-HK** [comment]
I've got it at 28 now so almost there, thanks Alex....I'm wondering if the way I been solving impulse issues is causing the damage to be off....we need to double check Manta and Raptor again

**Comment by HellDragon-HK** [comment]
@GreatEmerald If we're trying the UT3 Drawscales and we know the turrets are going to be an issue, would it be possible to have an alternate Levi that keeps the turrets where they are instead of using the extend animation and either having it check the map it's on to decide on it's own what to use from like a black list (if that's even possible) or have it as an extra mutator for use with smaller maps like Urban

**Comment by HellDragon-HK** [comment]
Added the link for UT3 Drawscale ukx edit

**Comment by GreatEmerald** [comment]
On UT3Weapons{,2}: are those just .uax files? If so, there's no problem. There should be no reliance on UT3Weapons.u, but .uax is fine.

On particles (beams etc.): they should be ported from UT3 as well. It's not yet done but it should for the most part be possible to port it truly to the original too. No sense in modifying the placeholder particle emitters in the mean while.

On smoke and fire: we keep to UT3 values, and they are in fact listed right there in the code: http://greatemerald.xmpcommunity.com/uncodex/UT3-15/Source_utgamecontent/utvehicle_leviathan_content.html#475 0.8 and 0.9.

Like I mentioned, an extra mutator that applies per-map fixes would be fine, but not very high priority because it's literally a hack.

**Comment by HellDragon-HK** [comment]
Being a sound yes I'm fairly sure they are uax so no need to pay attention to those then.

So you really do want it to be on fire with 80% health remaining, I still think that was a bug where they meant 10% and 20%, at the 90% and 80% you can't tell if an enemy Levi is actually truly damaged, it could be near full health and on fire which doesn't make sense to me.

You meant editing the maps themselves though, I meant like a MapEval type mod where it could be check marked by the user if it would use the alternate....which actually may not be needed after all there is more room in Urban than I though even with the full scaled main cannon it didn't go through the roof (although it probably can if you deploy in certain spots)

**Comment by HellDragon-HK** [comment]
Collision updated, fixing the suspensions caused a lot of movement trouble to the point I had to really make the collision on it tight, the lowest ladder panel can clip now but it's unavoidable....something I would like to ask, there is a collision height line in the class, does that double the collision box, although it doesn't seem to from the ground perspective but is it making it higher?

**Comment by HellDragon-HK** [comment]
@GreatEmerald 
In this video at around 29:47, is it normal for the Levi to make that kind of shot?
https://www.youtube.com/watch?v=Q6juQeJt9sU

**Comment by GreatEmerald** [comment]
Hm, that's true, they might have meant 10% and 20% for when the fire and smoke should start... Does the Leviathan go ablaze that easily in UT3, really? On the other hand, making it go ablaze easily is a way to make sure that it respawns faster if abandoned. What's your opinion on that, @Unre-Alex and @100GPing100?

And no, I meant adding a mutator that would do map-specific checks and adjust parameters if it detects a particular map.

I'm not sure how the collision affects it (I assume you mean `CollisionHeight`?). It would be nice to know for reference.

Yes, I don't see an issue with that shot?..

**Comment by GreatEmerald** [comment]
Also, here too you should split issues relating to each turret type into their own issues.

**Comment by HellDragon-HK** [comment]
Vehicles in UT3 respawn very very quickly if you walk away even just slightly and the Leviathan is worse than all of them in this regard, on trying to smack it with a Raptor just before I collided it respawned right in front of me for example.....plus it can't be hijacked at all

Yes it really does go ablaze at very early health in both PC & PS3

So a single shot too a 100% power core and killing it is normal then? Ok then, guess it's just new to me.

The Collision Height and Radius to me still feel like they do nothing and I still believe are overridden by the boxes.

**Comment by HellDragon-HK** [comment]
@GreatEmerald While looking at the death messages for UT3 (now that I know where they are), I found this which is more proof and identifies the Levi's crush ability

[UTDmgType_LeviathanCollision]
DeathString="`k's Leviathan barrelled through `o."
MaleSuicide="`o had an accident."
FemaleSuicide="`o had an accident."

**Comment by GreatEmerald** [comment]
Hm, I think it should take 3 shots to destroy a power core. You can of course test that in UT3. But I think the cores have 4500 HP by default, and the blast should deal about 250*6=1500 points of damage.

Anyway, that point is in fact about a weapon, and not the Leviathan vehicle class itself. You should split the main cannon points to its own issue as well. Otherwise this issue looks good, much more manageable now!

Hm, the collision messages could be for running over pedestrians, not necessarily vehicles. Not saying that it doesn't do it, mind you, I never tested it myself.

**Comment by HellDragon-HK** [comment]
4500 in both games though?

Ok is this it for UT2004?
    ````DamageCapacity=4500````

If so uhh UT3

````DamageCapacity=5000````

**Comment by HellDragon-HK** [comment]
@GreatEmerald Wait here it is I found it only when viewing source code
http://greatemerald.xmpcommunity.com/uncodex/UT3-15/Source_utgame/utvehicle_leviathan.html#30 line 0938

**Comment by HellDragon-HK** [comment]
Animation file was updated to scale up the driver turret, just like the Paladin the turret size is much larger than the vehicle's drawscale apparently....anybody know why because I haven't seen any turret specific drawscales in the UT3 code

**Comment by GreatEmerald** [comment]
I was talking about a UT3 Leviathan shooting at a UT2004 power core. Yes, power cores take more damage in UT3.

Interesting. Unfortunately the code for crushing pedestrians is implemented natively, so we can't reuse that hook. I guess it might be possible to simulate this effect with impact events like the Scorpion self-destruct does.

That would be very weird. The DrawScale of the turret should always match the DrawScale of the vehicle, because the way 100GPing100 ported it is simply by separating the turrets into their own meshes. Maybe you scaled up the vehicle mesh and didn't scale up the turrets by the same amount?

**Comment by HellDragon-HK** [comment]
I was talking about the main cannon hitting a power core yes and taking it out with a single hit yes, still does currently.

That crush is for vehicles specifically though isn't it, normal roadkill I think still applies.....ah death messages actually worked in my game it wouldn't be a problem to find out :( 

Ee found this issue with the Paladin turret being made smaller by a 0.8 in the stock UT2004 code, probably a smaller scale in the stock Levi turret too, haven't had the chance to check that yet.

**Comment by GreatEmerald** [comment]
Yes, it is. Ideally the UT3 code would be reusable as-is, just would need to be put into the impact function.

**Comment by HellDragon-HK** [comment]
Found what was causing the small turret, updated ukx to have 1.25 like it should and updated pull request to fix the issue which is in UT3LeviathanDriverWeapon.uc, change drawscale from 0.6 to 1.0
